Update SSH keys via MCD

kikisdeliveryservice opened the pull request:
Allow MCD to update SSH Keys.  

We are only allowing one user named "core" and only updating the field SSHAuthorizedKeys for this user.

kikisdeliveryservice commented:
/assign @ashcrow

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -100,17 +104,58 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) (bool
+					break
+				}
+			}
+			// length of Users is different, we don't do those kinds of changes, so nope.
```
I'm stuck between an attempted addition of a new user should noop (as it's doing here) or if we should raise the irreconcilable flag as the user was attempting to do something we are not going to do. @sdemos thoughts?

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -100,17 +104,58 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) (bool
+		// Passwd configs is an SSHAuthorizedKey change
+		// Any change other than an SSHChanage will return false.
+
+		var onlySSHChanged bool
```
nit: It is true that bools are false by default ingolang, but `onlySSHChanged := false` would be a bit more clear for a casual reader.

kikisdeliveryservice replied:
Done.

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -100,17 +104,58 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) (bool
+			}
+			// length of Users is different, we don't do those kinds of changes, so nope.
+		} else {
+			// prob unnecessary since defaul bool value is false
```
Agreed. This else can be dropped, though it's not a review blocker.

kikisdeliveryservice replied:
Done.

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -176,6 +221,19 @@ func (dn *Daemon) updateFiles(oldConfig, newConfig *mcfgv1.MachineConfig) error
 	return nil
 }
 
+//updatePasswd can be used to update Passwd with changes for Users at indices contained
```
nit: missing space after comment

kikisdeliveryservice replied:
Done.

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -406,6 +464,22 @@ func (dn *Daemon) writeFiles(files []ignv2_2types.File) error {
+			// ssh keys are in PasswdUser.HomeDir/.ssh
+			path = filepath.Join(oldUsers[idx].HomeDir, ".ssh")
+			glog.V(2).Infof("Writing SSHKeys at %q:", path)
+			// write the SSHKeys
```
:+1:

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -406,6 +464,22 @@ func (dn *Daemon) writeFiles(files []ignv2_2types.File) error {
+			glog.V(2).Infof("Writing SSHKeys at %q:", path)
+			// write the SSHKeys
+		} else {
+			return fmt.Errorf("Failed to write SSHKeys")
```
Once the ssh key writing is added I recommend returning the error that is bubbled up from that work OR adding a log statement with details before this return

kikisdeliveryservice commented:
Thanks @ashcrow for your thoughts!  I'll incorporate them into the PR. :smiley_cat:

smarterclayton commented:
/test e2e-aws

ericavonb commented:
/cc @openshift/sig-security

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -470,15 +470,31 @@ func (dn *Daemon) writeFiles(files []ignv2_2types.File) error {
+
+			var concatSSHKeys string
+			for _, k := range newUsers[idx].SSHAuthorizedKeys {
+				concatSSHKeys = concatSSHKeys + string(k) + "/n"
```
Why `/n` and not `\n`?

kikisdeliveryservice replied:
DOH

kikisdeliveryservice replied:
@rphillips Thank you for spotting that!

kikisdeliveryservice replied:
@rphillips ah just remembered that I tried strings.Join before, but it won't work bc SSHAuthorizedKeys -> type []SSHAuthorizedKey -> type SSHAuthorizedKey string

kikisdeliveryservice commented on pkg/daemon/update.go:
```diff
@@ -470,15 +470,31 @@ func (dn *Daemon) writeFiles(files []ignv2_2types.File) error {
+				concatSSHKeys = concatSSHKeys + string(k) + "/n"
+			}
+
+			err := ioutil.WriteFile(authkeypath, []byte(concatSSHKeys), os.FileMode(0600))
```
That seems like it would make sense.  The other funcs in update.go are also using ioutils.WriteFile  so it will be helpful. :)

kikisdeliveryservice commented:
/retest

ashcrow commented:
/retest

Flakes.

kikisdeliveryservice commented:
Temporarily blocked by:
https://github.com/openshift/installer/issues/578

ashcrow commented:
Woo! All green!

ashcrow commented:
Rebase needed.

eparis commented:
Are we going to be able to get back to this soon? I'm in a fight with @crawford here: https://github.com/openshift/installer/issues/862
I'm not happy shipping without ssh keys without this PR...

ashcrow commented:
Yes. Kirsten is currently at Kubecon but will be coming back to this soon.

cgwalters commented on pkg/daemon/update.go:
```diff
@@ -121,17 +129,54 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) (bool
+
+		sshReconcilable := false
+		// If the size of list Users is different, a User has been added or deleted -> nope
+		if len(oldIgn.Passwd.Users) == len(newIgn.Passwd.Users) {
```
Probably worth splitting this into a separate `sshConfigsReconcilable()` or so.

kikisdeliveryservice replied:
This is subject to change once my other PR lands.

cgwalters commented on pkg/daemon/update_test.go:
```diff
@@ -63,28 +62,29 @@ func TestUpdateOS(t *testing.T) {
 }
 
+// checkReconcilableResults is a shortcut for verifying results that should be reconcilable
+func checkReconcilableResults(t *testing.T, key string, err error, isReconcilable bool) {
```
Could split up these test changes and land them now.

kikisdeliveryservice commented:
Reported the tests that are failing due to "cluster-osin-operator". Awaiting new payload getting built.

kikisdeliveryservice commented:
/retest

kikisdeliveryservice commented:
/retest

ashcrow commented:
/test e2e-aws

kikisdeliveryservice commented:
Trying to get a cluster up to test this.

kikisdeliveryservice commented:
/approve cancel

ashcrow commented:
Let me know when I should review!

kikisdeliveryservice commented:
Ran into an Ignition config issue in testing where new configs are appended which hurts the current reconcilable logic. Consulted with @abhinavdahiya and @ajeddeloh and will tweak reconcilable.

kikisdeliveryservice commented:
/hold

kikisdeliveryservice commented:
I'll pick this back up tomorrow, I'm still seeing some wonky ignition behaviour when it's combining the configs.  I'll update you tomorrow @ashcrow

kikisdeliveryservice commented:
/retest

ashcrow commented:
/test e2e-aws

ashcrow commented:
/test e2e-aws

kikisdeliveryservice commented:
Manual testing status:
- replace current key for user "core" with a new key via a machineconfig: Success!  Verified that I am able to ssh in with new key, can't ssh in with old key and authorized key file on worker is as predicted.

cgwalters commented on pkg/daemon/update.go:
```diff
@@ -262,6 +239,24 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) *stri
+			glog.Info("SSH Keys reconcilable")
+		}
+	} else {
+		msg = "Ignition passwd user section contains unsupported changes: user must be core and have 1 or more sshKeys"
```
Don't need to do in this PR, but it should be supported to entirely remove SSH keys I'd say.

kikisdeliveryservice replied:
I can double check on this req after this PR and follow-on, thanks!

cgwalters commented on pkg/daemon/update.go:
```diff
@@ -184,44 +184,21 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) *stri
 			// check if the prior config is empty and that this is the first time running.
 			// if so, the SSHKey from the cluster config and user "core" must be added to machine config.
 			if len(oldIgn.Passwd.Users) == 0 && len(newIgn.Passwd.Users) == 1 {
```
Offhand, it feels like this "first time running" is actually just for transitioning clusters from *before* this PR to clusters with it (or in the future to support re-installing SSH keys if they've been removed).  Freshly installed clusters *with* this PR and a SSH key supplied should always have `len(oldIgn.Passwd.Users)` right?

kikisdeliveryservice replied:
Nope the first time running part is used from my first SSH PR where we are passing the SSH Keys from the cluster config thru to the MCD.

kikisdeliveryservice replied:
I cleaned up this section in #293

cgwalters commented on pkg/daemon/update.go:
```diff
@@ -184,44 +184,21 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) *stri
-						msg := "Ignition passwd user section contains unsupported changes"
+			} else if len(oldIgn.Passwd.Users) > 0 && len(newIgn.Passwd.Users) >= 1 {
+				for _, user := range newIgn.Passwd.Users {
+					if user.Name != "core" {
```
It'd probably be clearer to explicitly error out if `len(newIgn.Passwd.Users) > 1` since we only support the "core" user, and there can only be one user with a given name so something like:

```
if len(newIgn.Passwd.Users) > 1 {
   msg := "Adding multiple users is not supported"
   return &msg
}
```

If we do that *first*, every following conditional can then be sure the length is either 0 or 1.

cgwalters replied:
Basically all the code after that could actually operate on a `*User` which can be `nil` (in Rust terms `Option<&User>`) and we don't need to think about arrays of users.

IOW after the above code, something like:

```
newUser := nil
oldUser := nil
if len(newIgn.Passwd.Users) > 0 {
    newUser = &newIgn.Passwd.Users[0];
}
if len(newIgn.Passwd.Users) > 0 {
    oldUser = &oldIgn.Passwd.Users[0];
}
```

And then instead of doing `newUsers[0].Name` it'd just be `newUser.Name`.

kikisdeliveryservice replied:
You would think this would work and this suggestion is closer to how this PR was initially structured _before_ I discovered Ignition's quirks :sob:  

Short explanation: the way Ignition works the newUsers array is actually the compilation of all the Users from all of the machineconfigs in `oc get machineconfigs`. What ignition does is only _apply_ the last one. So that the newUsers array will likely always be greater than 1, except for the first time this runs when getting the initial SSH Key from the installer.

cgwalters commented:
(Though to be clear I'm also OK to `lgtm` these changes as is, particularly since they've been tested; can always do follow up work later)

cgwalters commented:
/approve

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -180,20 +180,30 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) *stri
 		if !reflect.DeepEqual(oldIgn.Passwd.Users, newIgn.Passwd.Users) {
+			// check if the prior config is empty and that this is the first time running.
+			// if so, the SSHKey from the cluster config and user "core" must be added to machine config.
+			glog.Infof("oldUsers: %v", oldIgn.Passwd.Users)
```
Thank you for logging.

kikisdeliveryservice replied:
retained the logging that made most sense looking at my cluster. we can add more/less in a follow-on if necessary :)

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -233,6 +243,26 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) *stri
 	return nil
 }
 
+func verifyUserFields(pwdUser ignv2_2types.PasswdUser) string {
```
If you end up making changes to this PR please add a doc to this function. While it's not required being an internal function, it's still helpful for future developers.

kikisdeliveryservice replied:
Done.

kikisdeliveryservice commented:
/hold

kikisdeliveryservice commented:
/hold cancel

ashcrow commented:
/retest

kikisdeliveryservice commented:
Passed manual testing, only changes made since last review:
- removed some of my logging that was added for testing
- added additional comments for clarity
- squashed commits

cgwalters commented on pkg/daemon/update.go:
```diff
@@ -173,27 +173,38 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) *stri
+					}
+				}
+				glog.Infof("user data to be verified before ssh update: %v", newIgn.Passwd.Users[len(newIgn.Passwd.Users)-1])
+				msg := verifyUserFields(newIgn.Passwd.Users[len(newIgn.Passwd.Users)-1])
```
Should this be `verifyUserFields(user)`?

kikisdeliveryservice replied:
no, we are taking the last user in users in line with ignition logic.

cgwalters replied:
Wait sorry, I missed this was outside the `for` loop.

ashcrow commented on pkg/daemon/update.go:
```diff
@@ -233,6 +244,29 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) *stri
 	return nil
 }
 
+// verifyUserFields returns nil if the user Name = "core", if 1 or more SSHKeys exist for
```
Thank you!

cgwalters commented:
/lgtm

ashcrow commented:
/retest

ashcrow commented:
Looks like @cgwalters reviewed at the same time.

/lgtm

ashcrow commented:
Infra failure

```
could not wait for build: the build machine-config-controller failed with reason PullBuilderImageFailed: Failed pulling builder image
```

ashcrow commented:
/test images

ashcrow commented:
/retest

sjenning commented:
We are hitting the same image build failures in https://github.com/openshift/machine-config-operator/pull/249

ashcrow commented:
/retest

ashcrow commented:
/retest

kikisdeliveryservice commented:
/retest

kikisdeliveryservice commented:
/retest

kikisdeliveryservice commented:
/retest

kikisdeliveryservice commented:
/retest

kikisdeliveryservice commented:
/retest

kikisdeliveryservice commented:
/test e2e-aws

kikisdeliveryservice commented:
My tests finally all passed and now they are rerunning. :frowning_face:

ashcrow commented:
CI MERGED! :tada:
